Add ESP-IDF (Ethernet) support by abstracting UDP transport (AsyncUDP → BSD sockets)

# Add ESP-IDF (Ethernet) support by abstracting UDP transport (AsyncUDP → BSD sockets)

**Type:** Feature / Refactor
**Area:** Networking, Transport, Build system
**Goal:** Make the gateway work over *wired Ethernet* on ESP32‑C3 (ESP‑IDF) in addition to the current Wi‑Fi (Arduino core) path.

---

## Summary

The project currently depends on **AsyncUDP** (Arduino core) for UDP I/O. To support **ESP‑IDF + Ethernet (DM9051)** on ESP32‑C3, we will:

1. **Abstract the UDP transport** used by `NibeGwComponent`.
2. Provide two backends:

   * **Arduino:** keep `AsyncUDP` (unchanged for existing users).
   * **ESP‑IDF:** implement a small **BSD sockets (lwIP)** shim (`socket/bind/recvfrom/sendto`).
3. **Remove Wi‑Fi‑only includes** and **avoid adding AsyncUDP libs** when building for ESP‑IDF.
4. *(Optional but recommended)* Replace Arduino types (`byte`, `boolean`) with standard types (`uint8_t`, `bool`).

The RS‑485 framing/state machine and climate logic remain unchanged.

---

## Motivation

* ESP32‑C3 wired Ethernet (DM9051) requires **ESP‑IDF** framework.
* `AsyncUDP` is Arduino‑only; it fails on IDF builds.
* ESPHome’s `network::is_connected()` already abstracts link state (Wi‑Fi or ETH), so only the UDP transport layer needs to change.

---

## Scope

### In‑scope

* Introduce a dual‑backend UDP transport in `NibeGwComponent` with compile‑time selection.
* Adjust includes and build flags so ESP‑IDF builds don’t pull Arduino/AsyncUDP.
* Maintain backward compatibility for existing Arduino (Wi‑Fi) users.

### Out‑of‑scope

* RS‑485/Nibe protocol handling.
* Climate entities or user‑facing behavior beyond enabling wired networking.

---

## Implementation Plan

### 1) Introduce transport abstraction (preprocessor‑guarded)

**`NibeGwComponent.h`**

* Remove Wi‑Fi headers for ESP32 builds.
* Keep `AsyncUDP` only when compiling for Arduino.
* For ESP‑IDF, add minimal lwIP includes and replace `AsyncUDP` members with socket FDs.

**Proposed diff (illustrative):**

```diff
-#ifdef USE_ESP32
-#include <WiFi.h>
-#include "AsyncUDP.h"
-#endif
+#if defined(ARDUINO)
+  #include "AsyncUDP.h"        // Arduino core path
+#else
+  // ESP-IDF path (lwIP / BSD sockets)
+  #include "lwip/sockets.h"
+  #include "lwip/inet.h"
+  #include "lwip/ip_addr.h"
+  #include <fcntl.h>
+#endif

   class NibeGwComponent : public Component {
     ...
-    AsyncUDP udp_read_;
-    AsyncUDP udp_write_;
+  #if defined(ARDUINO)
+    AsyncUDP udp_read_;
+    AsyncUDP udp_write_;
+  #else
+    int udp_read_fd_  = -1;
+    int udp_write_fd_ = -1;
+  #endif
     bool is_connected_{false};
```

### 2) Dual‑backend behavior in `NibeGwComponent.cpp`

* **Connect/Disconnect:**

  * Arduino: `udp.listen(port)` / `.close()` (unchanged)
  * IDF: `socket()` + `bind()` (read FD), `socket()` (write FD), non‑blocking via `fcntl`, and `close()` on disconnect.

* **Receive:**

  * Arduino: existing `.onPacket(...)` callbacks (unchanged)
  * IDF: poll `recvfrom()` inside `loop()`; for each packet, enqueue just like the Arduino path.

* **Send:**

  * Arduino: existing `.writeTo(...)` (unchanged)
  * IDF: `sendto()` for each configured target.

**Proposed diff (key excerpts):**

```diff
@@ void NibeGwComponent::loop() {
-  if (network::is_connected() && !is_connected_) {
-    udp_read_.listen(udp_read_port_);
-    udp_write_.listen(udp_write_port_);
-    is_connected_ = true;
-  }
+  if (network::is_connected() && !is_connected_) {
+  #if defined(ARDUINO)
+    udp_read_.listen(udp_read_port_);
+    udp_write_.listen(udp_write_port_);
+  #else
+    // read socket
+    udp_read_fd_ = ::socket(AF_INET, SOCK_DGRAM, IPPROTO_UDP);
+    if (udp_read_fd_ >= 0) {
+      fcntl(udp_read_fd_, F_SETFL, O_NONBLOCK);
+      sockaddr_in a{}; a.sin_family = AF_INET; a.sin_port = htons(udp_read_port_); a.sin_addr.s_addr = INADDR_ANY;
+      bind(udp_read_fd_, (sockaddr*)&a, sizeof(a));
+    }
+    // write socket
+    udp_write_fd_ = ::socket(AF_INET, SOCK_DGRAM, IPPROTO_UDP);
+    if (udp_write_fd_ >= 0) fcntl(udp_write_fd_, F_SETFL, O_NONBLOCK);
+  #endif
+    is_connected_ = true;
+  }

@@ void NibeGwComponent::loop() {
-  // existing logic...
+  // existing logic...
+  // IDF receive pump
+#if !defined(ARDUINO)
+  if (is_connected_ && udp_read_fd_ >= 0) {
+    uint8_t buf[MAX_DATA_LEN];
+    sockaddr_in from{}; socklen_t flen = sizeof(from);
+    int n = recvfrom(udp_read_fd_, buf, sizeof(buf), 0, (sockaddr*)&from, &flen);
+    while (n > 0) {
+      uint32_t host_ip = ntohl(from.sin_addr.s_addr);
+      network::IPAddress ip(host_ip);
+      if (source_ip_allowed_(ip)) {
+        handle_udp_read_packet_(buf, n, ip); // mirrors AsyncUDP callback
+      }
+      n = recvfrom(udp_read_fd_, buf, sizeof(buf), 0, (sockaddr*)&from, &flen);
+    }
+  }
+#endif

@@ void NibeGwComponent::send_udp_(const uint8_t* data, size_t len) {
-  for (auto &t : udp_targets_) {
-    ip_addr_t addr = (ip_addr_t) std::get<0>(t);
-    udp_read_.writeTo(data, len, &addr, std::get<1>(t));
-  }
+  for (auto &t : udp_targets_) {
+  #if defined(ARDUINO)
+    ip_addr_t addr = (ip_addr_t) std::get<0>(t);
+    udp_read_.writeTo(data, len, &addr, std::get<1>(t));
+  #else
+    sockaddr_in to{}; to.sin_family = AF_INET;
+    to.sin_port = htons(std::get<1>(t));
+    to.sin_addr.s_addr = htonl((uint32_t) std::get<0>(t));
+    sendto(udp_write_fd_, data, len, 0, (sockaddr*)&to, sizeof(to));
+  #endif
+  }
```

> **Note:** Add a small helper `handle_udp_read_packet_(const uint8_t* data, int len, const network::IPAddress& from)` that contains the exact queuing logic currently done inside the `AsyncUDPPacket` callback. The Arduino path calls it from `.onPacket(...)`; the IDF path calls it from the `recvfrom()` pump.

* **Disconnect:**

```diff
-  udp_read_.close();
-  udp_write_.close();
+ #if defined(ARDUINO)
+   udp_read_.close();
+   udp_write_.close();
+ #else
+   if (udp_read_fd_  >= 0) { ::close(udp_read_fd_);  udp_read_fd_  = -1; }
+   if (udp_write_fd_ >= 0) { ::close(udp_write_fd_); udp_write_fd_ = -1; }
+ #endif
   is_connected_ = false;
```

### 3) Guard library additions in `__init__.py`

* Only add the AsyncUDP library when building for **Arduino**.
* For ESP‑IDF builds, **don’t** add AsyncUDP (the sockets backend uses the IDF toolchain).

```diff
- if CORE.is_esp32:
-     cg.add_build_flag("-DHARDWARE_SERIAL_WITH_PINS")
-     cg.add_library("ESP32 Async UDP", None)
+ if CORE.is_esp32:
+     cg.add_build_flag("-DHARDWARE_SERIAL_WITH_PINS")
+     # AsyncUDP is Arduino-only; do not add for ESP-IDF
+     if getattr(CORE, "using_arduino", False):
+         cg.add_library("ESP32 Async UDP", None)
```

(If the ESPHome version doesn’t expose `CORE.using_arduino`, simply omit the `cg.add_library(...)` line; Arduino builds typically get AsyncUDP from the core/boards libs.)

### 4) *(Optional)* De‑Arduino‑fy base types in `NibeGw.h/.cpp`

Replace:

* `#include <Arduino.h>` → `#include <cstdint>`
* `byte` → `uint8_t`
* `boolean` → `bool`

This makes headers neutral and avoids accidental Arduino dependencies in the IDF build.

**Example (header):**

```diff
- #include <Arduino.h>
+ #include <cstdint>

- typedef std::function<void(const byte *const data, int len)> callback_msg_received_type;
+ using callback_msg_received_type = std::function<void(const uint8_t* data, int len)>;

- boolean connectionState;
+ bool connectionState;

- byte buffer[MAX_DATA_LEN * 2];
+ uint8_t buffer[MAX_DATA_LEN * 2];
```

---

## Acceptance Criteria

* Firmware builds and runs on **ESP32‑C3 + ESP‑IDF** with this YAML (example):

  ```yaml
  esp32:
    board: esp32-c3-devkitm-1
    framework:
      type: esp-idf

  ethernet:
    type: DM9051
    clk_pin: GPIO7
    mosi_pin: GPIO10
    miso_pin: GPIO3
    cs_pin: GPIO9
    interrupt_pin: GPIO8
    reset_pin: GPIO6
    clock_speed: 8MHz
    manual_ip:
      static_ip: 192.168.xx.yy
      gateway: 192.168.xx.1
      subnet: 255.255.255.0
      dns1: 192.168.xx.1
    use_address: 192.168.xx.yy
  ```

* Existing **Arduino/Wi‑Fi** builds continue to work without YAML or code changes.

* UDP RX/TX behaves identically for both backends (same queuing, same target delivery, source IP filtering still respected).

* No Wi‑Fi‑only headers or Arduino‑only libs used in ESP‑IDF builds.

I'm not sure how to force a redownload. I did a compile without the ref:  and then a clean and compile with ref: I hop that's enough.

```
INFO ESPHome 2025.11.4
INFO Reading configuration /config/esphome/heatpumpgw.yaml...
INFO Detected timezone 'Europe/Lisbon'
WARNING GPIO12 is a strapping PIN and should only be used for I/O with care.
Attaching external pullup/down resistors to strapping pins can cause unexpected failures.
See https://esphome.io/guides/faq.html#why-am-i-getting-a-warning-about-strapping-pins
INFO Generating C++ source...
INFO Compiling app... Build path: /data/build/heatpumpgw
Processing heatpumpgw (board: esp32dev; framework: espidf; platform: https://github.com/pioarduino/platform-espressif32/releases/download/55.03.31-2/platform-espressif32.zip)
--------------------------------------------------------------------------------
INFO Package configuration completed successfully
Library Manager: Installing esphome/noise-c @ 0.1.10
INFO Installing esphome/noise-c @ 0.1.10
Downloading  [####################################]  100%
Unpacking  [####################################]  100%
Library Manager: noise-c@0.1.10 has been installed!
INFO noise-c@0.1.10 has been installed!
Library Manager: Resolving dependencies...
INFO Resolving dependencies...
Library Manager: Installing esphome/libsodium @ 1.10020.7
INFO Installing esphome/libsodium @ 1.10020.7
Downloading  [####################################]  100%
Unpacking  [####################################]  100%
Library Manager: libsodium@1.10020.7 has been installed!
INFO libsodium@1.10020.7 has been installed!
INFO Package configuration completed successfully
HARDWARE: ESP32 240MHz, 320KB RAM, 4MB Flash
 - contrib-piohome @ 3.4.4 
 - framework-espidf @ 3.50501.0 (5.5.1) 
 - tool-cmake @ 4.0.3 
 - tool-esp-rom-elfs @ 2024.10.11 
 - tool-esptoolpy @ 5.1.0 
 - tool-mklittlefs @ 3.2.0 
 - tool-ninja @ 1.13.1 
 - tool-scons @ 4.40801.0 (4.8.1) 
 - toolchain-xtensa-esp-elf @ 14.2.0+20241119
Reading CMake configuration...
Generating assembly for certificate bundle...
Dependency Graph
|-- noise-c @ 0.1.10
Compiling .pioenvs/heatpumpgw/src/esphome/components/api/api_connection.cpp.o
Compiling .pioenvs/heatpumpgw/src/esphome/components/api/api_frame_helper.cpp.o
Compiling .pioenvs/heatpumpgw/src/esphome/components/api/api_frame_helper_noise.cpp.o
Compiling .pioenvs/heatpumpgw/src/esphome/components/api/api_pb2.cpp.o
Compiling .pioenvs/heatpumpgw/src/esphome/components/api/api_pb2_service.cpp.o
Compiling .pioenvs/heatpumpgw/src/esphome/components/api/api_server.cpp.o
Compiling .pioenvs/heatpumpgw/src/esphome/components/api/list_entities.cpp.o
Compiling .pioenvs/heatpumpgw/src/esphome/components/api/proto.cpp.o
Compiling .pioenvs/heatpumpgw/src/esphome/components/api/subscribe_state.cpp.o
Compiling .pioenvs/heatpumpgw/src/esphome/components/binary_sensor/automation.cpp.o
Compiling .pioenvs/heatpumpgw/src/esphome/components/binary_sensor/binary_sensor.cpp.o
Compiling .pioenvs/heatpumpgw/src/esphome/components/binary_sensor/filter.cpp.o
Compiling .pioenvs/heatpumpgw/src/esphome/components/button/button.cpp.o
Compiling .pioenvs/heatpumpgw/src/esphome/components/climate/climate.cpp.o
Compiling .pioenvs/heatpumpgw/src/esphome/components/climate/climate_mode.cpp.o
Compiling .pioenvs/heatpumpgw/src/esphome/components/climate/climate_traits.cpp.o
Compiling .pioenvs/heatpumpgw/src/esphome/components/esp32/core.cpp.o
Compiling .pioenvs/heatpumpgw/src/esphome/components/esp32/gpio.cpp.o
Compiling .pioenvs/heatpumpgw/src/esphome/components/esp32/helpers.cpp.o
Compiling .pioenvs/heatpumpgw/src/esphome/components/esp32/preferences.cpp.o
Compiling .pioenvs/heatpumpgw/src/esphome/components/esphome/ota/ota_esphome.cpp.o
Compiling .pioenvs/heatpumpgw/src/esphome/components/gpio/switch/gpio_switch.cpp.o
Compiling .pioenvs/heatpumpgw/src/esphome/components/homeassistant/time/homeassistant_time.cpp.o
Compiling .pioenvs/heatpumpgw/src/esphome/components/logger/logger.cpp.o
Compiling .pioenvs/heatpumpgw/src/esphome/components/logger/logger_esp32.cpp.o
Compiling .pioenvs/heatpumpgw/src/esphome/components/logger/task_log_buffer.cpp.o
Compiling .pioenvs/heatpumpgw/src/esphome/components/md5/md5.cpp.o
Compiling .pioenvs/heatpumpgw/src/esphome/components/mdns/mdns_component.cpp.o
Compiling .pioenvs/heatpumpgw/src/esphome/components/mdns/mdns_esp32.cpp.o
Compiling .pioenvs/heatpumpgw/src/esphome/components/network/util.cpp.o
Compiling .pioenvs/heatpumpgw/src/esphome/components/nibegw/NibeGw.cpp.o
Compiling .pioenvs/heatpumpgw/src/esphome/components/nibegw/NibeGwClimate.cpp.o
Compiling .pioenvs/heatpumpgw/src/esphome/components/nibegw/NibeGwComponent.cpp.o
Compiling .pioenvs/heatpumpgw/src/esphome/components/ota/ota_backend.cpp.o
src/esphome/components/nibegw/NibeGwComponent.cpp: In member function 'void esphome::nibegw::NibeGwComponent::recv_local_socket(std::unique_ptr<esphome::socket::Socket>&, int, int)':
src/esphome/components/nibegw/NibeGwComponent.cpp:76:45: error: no matching function for call to 'IPAddress(in_addr_t&)'
   76 |   network::IPAddress ip(from.sin_addr.s_addr);
      |                                             ^
In file included from src/esphome/components/nibegw/NibeGwComponent.h:14,
                 from src/esphome/components/nibegw/NibeGwComponent.cpp:1:
src/esphome/components/network/ip_address.h:90:3: note: candidate: 'esphome::network::IPAddress::IPAddress(esp_ip_addr_t*)' (near match)
   90 |   IPAddress(esp_ip_addr_t *other_ip) {
      |   ^~~~~~~~~
src/esphome/components/network/ip_address.h:90:3: note:   conversion of argument 1 would be ill-formed:
src/esphome/components/nibegw/NibeGwComponent.cpp:76:39: error: invalid conversion from 'in_addr_t' {aka 'long unsigned int'} to 'esp_ip_addr_t*' {aka '_ip_addr*'} [-fpermissive]
   76 |   network::IPAddress ip(from.sin_addr.s_addr);
      |                         ~~~~~~~~~~~~~~^~~~~~
      |                                       |
      |                                       in_addr_t {aka long unsigned int}
src/esphome/components/network/ip_address.h:84:3: note: candidate: 'esphome::network::IPAddress::IPAddress(esp_ip4_addr_t*)' (near match)
   84 |   IPAddress(esp_ip4_addr_t *other_ip) {
      |   ^~~~~~~~~
src/esphome/components/network/ip_address.h:84:3: note:   conversion of argument 1 would be ill-formed:
src/esphome/components/nibegw/NibeGwComponent.cpp:76:39: error: invalid conversion from 'in_addr_t' {aka 'long unsigned int'} to 'esp_ip4_addr_t*' {aka 'esp_ip4_addr*'} [-fpermissive]
   76 |   network::IPAddress ip(from.sin_addr.s_addr);
      |                         ~~~~~~~~~~~~~~^~~~~~
      |                                       |
      |                                       in_addr_t {aka long unsigned int}
src/esphome/components/network/ip_address.h:61:3: note: candidate: 'esphome::network::IPAddress::IPAddress(ip4_addr_t*)' (near match)
   61 |   IPAddress(ip4_addr_t *other_ip) {
      |   ^~~~~~~~~
src/esphome/components/network/ip_address.h:61:3: note:   conversion of argument 1 would be ill-formed:
src/esphome/components/nibegw/NibeGwComponent.cpp:76:39: error: invalid conversion from 'in_addr_t' {aka 'long unsigned int'} to 'ip4_addr_t*' {aka 'ip4_addr*'} [-fpermissive]
   76 |   network::IPAddress ip(from.sin_addr.s_addr);
      |                         ~~~~~~~~~~~~~~^~~~~~
      |                                       |
      |                                       in_addr_t {aka long unsigned int}
src/esphome/components/network/ip_address.h:59:3: note: candidate: 'esphome::network::IPAddress::IPAddress(const char*)' (near match)
   59 |   IPAddress(const char *in_address) { ipaddr_aton(in_address, &ip_addr_); }
      |   ^~~~~~~~~
src/esphome/components/network/ip_address.h:59:3: note:   conversion of argument 1 would be ill-formed:
src/esphome/components/nibegw/NibeGwComponent.cpp:76:39: error: invalid conversion from 'in_addr_t' {aka 'long unsigned int'} to 'const char*' [-fpermissive]
   76 |   network::IPAddress ip(from.sin_addr.s_addr);
      |                         ~~~~~~~~~~~~~~^~~~~~
      |                                       |
      |                                       in_addr_t {aka long unsigned int}
src/esphome/components/network/ip_address.h:58:3: note: candidate: 'esphome::network::IPAddress::IPAddress(const ip_addr_t*)' (near match)
   58 |   IPAddress(const ip_addr_t *other_ip) { ip_addr_copy(ip_addr_, *other_ip); }
      |   ^~~~~~~~~
src/esphome/components/network/ip_address.h:58:3: note:   conversion of argument 1 would be ill-formed:
src/esphome/components/nibegw/NibeGwComponent.cpp:76:39: error: invalid conversion from 'in_addr_t' {aka 'long unsigned int'} to 'const ip_addr_t*' {aka 'const ip4_addr*'} [-fpermissive]
   76 |   network::IPAddress ip(from.sin_addr.s_addr);
      |                         ~~~~~~~~~~~~~~^~~~~~
      |                                       |
      |                                       in_addr_t {aka long unsigned int}
src/esphome/components/network/ip_address.h:43:8: note: candidate: 'constexpr esphome::network::IPAddress::IPAddress(const esphome::network::IPAddress&)' (near match)
   43 | struct IPAddress {
      |        ^~~~~~~~~
src/esphome/components/network/ip_address.h:43:8: note:   conversion of argument 1 would be ill-formed:
src/esphome/components/nibegw/NibeGwComponent.cpp:76:39: error: conversion from 'in_addr_t' {aka 'long unsigned int'} to 'const esphome::network::IPAddress' is ambiguous
   76 |   network::IPAddress ip(from.sin_addr.s_addr);
      |                         ~~~~~~~~~~~~~~^~~~~~
src/esphome/components/network/ip_address.h:90:3: note: candidate: 'esphome::network::IPAddress::IPAddress(esp_ip_addr_t*)' (near match)
   90 |   IPAddress(esp_ip_addr_t *other_ip) {
      |   ^~~~~~~~~
src/esphome/components/network/ip_address.h:90:3: note:   conversion of argument 1 would be ill-formed:
src/esphome/components/nibegw/NibeGwComponent.cpp:76:39: error: invalid conversion from 'in_addr_t' {aka 'long unsigned int'} to 'esp_ip_addr_t*' {aka '_ip_addr*'} [-fpermissive]
   76 |   network::IPAddress ip(from.sin_addr.s_addr);
      |                         ~~~~~~~~~~~~~~^~~~~~
      |                                       |
      |                                       in_addr_t {aka long unsigned int}
src/esphome/components/network/ip_address.h:84:3: note: candidate: 'esphome::network::IPAddress::IPAddress(esp_ip4_addr_t*)' (near match)
   84 |   IPAddress(esp_ip4_addr_t *other_ip) {
      |   ^~~~~~~~~
src/esphome/components/network/ip_address.h:84:3: note:   conversion of argument 1 would be ill-formed:
src/esphome/components/nibegw/NibeGwComponent.cpp:76:39: error: invalid conversion from 'in_addr_t' {aka 'long unsigned int'} to 'esp_ip4_addr_t*' {aka 'esp_ip4_addr*'} [-fpermissive]
   76 |   network::IPAddress ip(from.sin_addr.s_addr);
      |                         ~~~~~~~~~~~~~~^~~~~~
      |                                       |
      |                                       in_addr_t {aka long unsigned int}
src/esphome/components/network/ip_address.h:61:3: note: candidate: 'esphome::network::IPAddress::IPAddress(ip4_addr_t*)' (near match)
   61 |   IPAddress(ip4_addr_t *other_ip) {
      |   ^~~~~~~~~
src/esphome/components/network/ip_address.h:61:3: note:   conversion of argument 1 would be ill-formed:
src/esphome/components/nibegw/NibeGwComponent.cpp:76:39: error: invalid conversion from 'in_addr_t' {aka 'long unsigned int'} to 'ip4_addr_t*' {aka 'ip4_addr*'} [-fpermissive]
   76 |   network::IPAddress ip(from.sin_addr.s_addr);
      |                         ~~~~~~~~~~~~~~^~~~~~
      |                                       |
      |                                       in_addr_t {aka long unsigned int}
src/esphome/components/network/ip_address.h:59:3: note: candidate: 'esphome::network::IPAddress::IPAddress(const char*)' (near match)
   59 |   IPAddress(const char *in_address) { ipaddr_aton(in_address, &ip_addr_); }
      |   ^~~~~~~~~
src/esphome/components/network/ip_address.h:59:3: note:   conversion of argument 1 would be ill-formed:
src/esphome/components/nibegw/NibeGwComponent.cpp:76:39: error: invalid conversion from 'in_addr_t' {aka 'long unsigned int'} to 'const char*' [-fpermissive]
   76 |   network::IPAddress ip(from.sin_addr.s_addr);
      |                         ~~~~~~~~~~~~~~^~~~~~
      |                                       |
      |                                       in_addr_t {aka long unsigned int}
src/esphome/components/network/ip_address.h:58:3: note: candidate: 'esphome::network::IPAddress::IPAddress(const ip_addr_t*)' (near match)
   58 |   IPAddress(const ip_addr_t *other_ip) { ip_addr_copy(ip_addr_, *other_ip); }
      |   ^~~~~~~~~
src/esphome/components/network/ip_address.h:58:3: note:   conversion of argument 1 would be ill-formed:
src/esphome/components/nibegw/NibeGwComponent.cpp:76:39: error: invalid conversion from 'in_addr_t' {aka 'long unsigned int'} to 'const ip_addr_t*' {aka 'const ip4_addr*'} [-fpermissive]
   76 |   network::IPAddress ip(from.sin_addr.s_addr);
      |                         ~~~~~~~~~~~~~~^~~~~~
      |                                       |
      |                                       in_addr_t {aka long unsigned int}
src/esphome/components/network/ip_address.h:60:3: note: candidate: 'esphome::network::IPAddress::IPAddress(const std::string&)'
   60 |   IPAddress(const std::string &in_address) { ipaddr_aton(in_address.c_str(), &ip_addr_); }
      |   ^~~~~~~~~
src/esphome/components/network/ip_address.h:60:32: note:   no known conversion for argument 1 from 'in_addr_t' {aka 'long unsigned int'} to 'const std::string&' {aka 'const std::__cxx11::basic_string<char>&'}
   60 |   IPAddress(const std::string &in_address) { ipaddr_aton(in_address.c_str(), &ip_addr_); }
      |             ~~~~~~~~~~~~~~~~~~~^~~~~~~~~~
src/esphome/components/network/ip_address.h:55:3: note: candidate: 'esphome::network::IPAddress::IPAddress(uint8_t, uint8_t, uint8_t, uint8_t)'
   55 |   IPAddress(uint8_t first, uint8_t second, uint8_t third, uint8_t fourth) {
      |   ^~~~~~~~~
src/esphome/components/network/ip_address.h:55:3: note:   candidate expects 4 arguments, 1 provided
src/esphome/components/network/ip_address.h:54:3: note: candidate: 'esphome::network::IPAddress::IPAddress()'
   54 |   IPAddress() { ip_addr_set_zero(&ip_addr_); }
      |   ^~~~~~~~~
src/esphome/components/network/ip_address.h:54:3: note:   candidate expects 0 arguments, 1 provided
src/esphome/components/network/ip_address.h:43:8: note: candidate: 'constexpr esphome::network::IPAddress::IPAddress(const esphome::network::IPAddress&)'
   43 | struct IPAddress {
      |        ^~~~~~~~~
src/esphome/components/network/ip_address.h:43:8: note:   no known conversion for argument 1 from 'in_addr_t' {aka 'long unsigned int'} to 'const esphome::network::IPAddress&'
src/esphome/components/network/ip_address.h:43:8: note: candidate: 'constexpr esphome::network::IPAddress::IPAddress(esphome::network::IPAddress&&)'
src/esphome/components/network/ip_address.h:43:8: note:   no known conversion for argument 1 from 'in_addr_t' {aka 'long unsigned int'} to 'esphome::network::IPAddress&&'
src/esphome/components/network/ip_address.h:43:8: note: candidate: 'constexpr esphome::network::IPAddress::IPAddress(esphome::network::IPAddress&&)' (near match)
src/esphome/components/network/ip_address.h:43:8: note:   conversion of argument 1 would be ill-formed:
src/esphome/components/nibegw/NibeGwComponent.cpp:76:39: error: conversion from 'in_addr_t' {aka 'long unsigned int'} to 'esphome::network::IPAddress' is ambiguous
   76 |   network::IPAddress ip(from.sin_addr.s_addr);
      |                         ~~~~~~~~~~~~~~^~~~~~
src/esphome/components/network/ip_address.h:90:3: note: candidate: 'esphome::network::IPAddress::IPAddress(esp_ip_addr_t*)' (near match)
   90 |   IPAddress(esp_ip_addr_t *other_ip) {
      |   ^~~~~~~~~
src/esphome/components/network/ip_address.h:90:3: note:   conversion of argument 1 would be ill-formed:
src/esphome/components/nibegw/NibeGwComponent.cpp:76:39: error: invalid conversion from 'in_addr_t' {aka 'long unsigned int'} to 'esp_ip_addr_t*' {aka '_ip_addr*'} [-fpermissive]
   76 |   network::IPAddress ip(from.sin_addr.s_addr);
      |                         ~~~~~~~~~~~~~~^~~~~~
      |                                       |
      |                                       in_addr_t {aka long unsigned int}
src/esphome/components/network/ip_address.h:84:3: note: candidate: 'esphome::network::IPAddress::IPAddress(esp_ip4_addr_t*)' (near match)
   84 |   IPAddress(esp_ip4_addr_t *other_ip) {
      |   ^~~~~~~~~
src/esphome/components/network/ip_address.h:84:3: note:   conversion of argument 1 would be ill-formed:
src/esphome/components/nibegw/NibeGwComponent.cpp:76:39: error: invalid conversion from 'in_addr_t' {aka 'long unsigned int'} to 'esp_ip4_addr_t*' {aka 'esp_ip4_addr*'} [-fpermissive]
   76 |   network::IPAddress ip(from.sin_addr.s_addr);
      |                         ~~~~~~~~~~~~~~^~~~~~
      |                                       |
      |                                       in_addr_t {aka long unsigned int}
src/esphome/components/network/ip_address.h:61:3: note: candidate: 'esphome::network::IPAddress::IPAddress(ip4_addr_t*)' (near match)
   61 |   IPAddress(ip4_addr_t *other_ip) {
      |   ^~~~~~~~~
src/esphome/components/network/ip_address.h:61:3: note:   conversion of argument 1 would be ill-formed:
src/esphome/components/nibegw/NibeGwComponent.cpp:76:39: error: invalid conversion from 'in_addr_t' {aka 'long unsigned int'} to 'ip4_addr_t*' {aka 'ip4_addr*'} [-fpermissive]
   76 |   network::IPAddress ip(from.sin_addr.s_addr);
      |                         ~~~~~~~~~~~~~~^~~~~~
      |                                       |
      |                                       in_addr_t {aka long unsigned int}
src/esphome/components/network/ip_address.h:59:3: note: candidate: 'esphome::network::IPAddress::IPAddress(const char*)' (near match)
   59 |   IPAddress(const char *in_address) { ipaddr_aton(in_address, &ip_addr_); }
      |   ^~~~~~~~~
src/esphome/components/network/ip_address.h:59:3: note:   conversion of argument 1 would be ill-formed:
src/esphome/components/nibegw/NibeGwComponent.cpp:76:39: error: invalid conversion from 'in_addr_t' {aka 'long unsigned int'} to 'const char*' [-fpermissive]
   76 |   network::IPAddress ip(from.sin_addr.s_addr);
      |                         ~~~~~~~~~~~~~~^~~~~~
      |                                       |
      |                                       in_addr_t {aka long unsigned int}
src/esphome/components/network/ip_address.h:58:3: note: candidate: 'esphome::network::IPAddress::IPAddress(const ip_addr_t*)' (near match)
   58 |   IPAddress(const ip_addr_t *other_ip) { ip_addr_copy(ip_addr_, *other_ip); }
      |   ^~~~~~~~~
src/esphome/components/network/ip_address.h:58:3: note:   conversion of argument 1 would be ill-formed:
src/esphome/components/nibegw/NibeGwComponent.cpp:76:39: error: invalid conversion from 'in_addr_t' {aka 'long unsigned int'} to 'const ip_addr_t*' {aka 'const ip4_addr*'} [-fpermissive]
   76 |   network::IPAddress ip(from.sin_addr.s_addr);
      |                         ~~~~~~~~~~~~~~^~~~~~
      |                                       |
      |                                       in_addr_t {aka long unsigned int}
src/esphome/components/network/ip_address.h:60:3: note: candidate: 'esphome::network::IPAddress::IPAddress(const std::string&)'
   60 |   IPAddress(const std::string &in_address) { ipaddr_aton(in_address.c_str(), &ip_addr_); }
      |   ^~~~~~~~~
src/esphome/components/network/ip_address.h:60:32: note:   no known conversion for argument 1 from 'in_addr_t' {aka 'long unsigned int'} to 'const std::string&' {aka 'const std::__cxx11::basic_string<char>&'}
   60 |   IPAddress(const std::string &in_address) { ipaddr_aton(in_address.c_str(), &ip_addr_); }
      |             ~~~~~~~~~~~~~~~~~~~^~~~~~~~~~
src/esphome/components/network/ip_address.h:55:3: note: candidate: 'esphome::network::IPAddress::IPAddress(uint8_t, uint8_t, uint8_t, uint8_t)'
   55 |   IPAddress(uint8_t first, uint8_t second, uint8_t third, uint8_t fourth) {
      |   ^~~~~~~~~
src/esphome/components/network/ip_address.h:55:3: note:   candidate expects 4 arguments, 1 provided
src/esphome/components/network/ip_address.h:54:3: note: candidate: 'esphome::network::IPAddress::IPAddress()'
   54 |   IPAddress() { ip_addr_set_zero(&ip_addr_); }
      |   ^~~~~~~~~
src/esphome/components/network/ip_address.h:54:3: note:   candidate expects 0 arguments, 1 provided
src/esphome/components/network/ip_address.h:43:8: note: candidate: 'constexpr esphome::network::IPAddress::IPAddress(const esphome::network::IPAddress&)'
   43 | struct IPAddress {
      |        ^~~~~~~~~
src/esphome/components/network/ip_address.h:43:8: note:   no known conversion for argument 1 from 'in_addr_t' {aka 'long unsigned int'} to 'const esphome::network::IPAddress&'
src/esphome/components/network/ip_address.h:43:8: note: candidate: 'constexpr esphome::network::IPAddress::IPAddress(esphome::network::IPAddress&&)'
src/esphome/components/network/ip_address.h:43:8: note:   no known conversion for argument 1 from 'in_addr_t' {aka 'long unsigned int'} to 'esphome::network::IPAddress&&'
src/esphome/components/network/ip_address.h:60:3: note: candidate: 'esphome::network::IPAddress::IPAddress(const std::string&)'
   60 |   IPAddress(const std::string &in_address) { ipaddr_aton(in_address.c_str(), &ip_addr_); }
      |   ^~~~~~~~~
src/esphome/components/network/ip_address.h:60:32: note:   no known conversion for argument 1 from 'in_addr_t' {aka 'long unsigned int'} to 'const std::string&' {aka 'const std::__cxx11::basic_string<char>&'}
   60 |   IPAddress(const std::string &in_address) { ipaddr_aton(in_address.c_str(), &ip_addr_); }
      |             ~~~~~~~~~~~~~~~~~~~^~~~~~~~~~
src/esphome/components/network/ip_address.h:55:3: note: candidate: 'esphome::network::IPAddress::IPAddress(uint8_t, uint8_t, uint8_t, uint8_t)'
   55 |   IPAddress(uint8_t first, uint8_t second, uint8_t third, uint8_t fourth) {
      |   ^~~~~~~~~
src/esphome/components/network/ip_address.h:55:3: note:   candidate expects 4 arguments, 1 provided
src/esphome/components/network/ip_address.h:54:3: note: candidate: 'esphome::network::IPAddress::IPAddress()'
   54 |   IPAddress() { ip_addr_set_zero(&ip_addr_); }
      |   ^~~~~~~~~
src/esphome/components/network/ip_address.h:54:3: note:   candidate expects 0 arguments, 1 provided
Compiling .pioenvs/heatpumpgw/src/esphome/components/ota/ota_backend_esp_idf.cpp.o
*** [.pioenvs/heatpumpgw/src/esphome/components/nibegw/NibeGwComponent.cpp.o] Error 1
========================= [FAILED] Took 36.27 seconds =========================

```

--- a/components/nibegw/NibeGw.h
+++ b/components/nibegw/NibeGw.h
@@ -36,15 +36,13 @@
 #ifndef NibeGw_h
 #define NibeGw_h
 
+#include <cstdint>
 #include "esphome/components/uart/uart.h"
 #include "esphome/core/gpio.h"
 #include <functional>
 #include <set>
 
 using namespace esphome;
-
-// #define HARDWARE_SERIAL_WITH_PINS
-// #define HARDWARE_SERIAL
 
 // state machine states
 enum eState {

--- a/components/nibegw/NibeGwComponent.cpp
+++ b/components/nibegw/NibeGwComponent.cpp
@@ -1,6 +1,7 @@
 #include "NibeGwComponent.h"
 
 namespace esphome {
+
 namespace nibegw {
 
 NibeGwComponent::NibeGwComponent(esphome::GPIOPin *dir_pin) {
@@ -9,9 +10,6 @@ NibeGwComponent::NibeGwComponent(esphome::GPIOPin *dir_pin) {
       std::bind(&NibeGwComponent::callback_msg_received, this, std::placeholders::_1, std::placeholders::_2),
       std::bind(&NibeGwComponent::callback_msg_token_received, this, std::placeholders::_1, std::placeholders::_2,
                 std::placeholders::_3));
-
-  udp_read_.onPacket([this](AsyncUDPPacket packet) { token_request_cache(packet, MODBUS40, READ_TOKEN); });
-  udp_write_.onPacket([this](AsyncUDPPacket packet) { token_request_cache(packet, MODBUS40, WRITE_TOKEN); });
 }
 
 static request_data_type dedup(const uint8_t *data, int len, uint8_t val) {
@@ -41,40 +39,43 @@ void NibeGwComponent::callback_msg_received(const uint8_t *data, int len) {
     return;
   }
 
-  for (auto target = udp_targets_.begin(); target != udp_targets_.end(); target++) {
-    ip_addr_t address = (ip_addr_t) std::get<0>(*target);
-    if (!udp_read_.writeTo(data, len, &address, std::get<1>(*target))) {
-      ESP_LOGW(TAG, "UDP Packet send failed to %s:%d", std::get<0>(*target).str().c_str(), std::get<1>(*target));
+  if (!udp_read_) {
+    ESP_LOGW(TAG, "UDP read socket not available");
+    return;
+  }
+
+  // Send to all UDP targets
+  for (auto &&target : udp_targets_) {
+    int result = udp_read_->sendto(data, len, 0, (sockaddr *) &target.storage, target.len);
+    if (result < 0) {
+      ESP_LOGW(TAG, "UDP sendto failed to %s, error: %d", target.str().c_str(), errno);
     }
   }
 }
 
-void NibeGwComponent::token_request_cache(AsyncUDPPacket &udp, uint8_t address, uint8_t token) {
-  if (!is_connected_) {
+void NibeGwComponent::recv_local_socket(std::unique_ptr<socket::Socket> &fd, int address, int token) {
+  request_data_type request(MAX_DATA_LEN);
+
+  socket_address from;
+  int n = fd->recvfrom(request.data(), request.size(), (sockaddr *) &from.storage, &from.len);
+  if (n < 0) {
+    if (errno != EAGAIN && errno != EWOULDBLOCK) {
+      ESP_LOGW(TAG, "recvfrom error on read socket: %d", errno);
+    }
     return;
   }
+  request.resize(n);
 
-  int size = udp.length();
-  if (size == 0) {
+  if (udp_sources_.size() &&
+      none_of(udp_sources_.begin(), udp_sources_.end(), [&](auto &source) { return from.matches(source); })) {
+    ESP_LOGW(TAG, "UDP Packet wrong ip ignored %s", from.str().c_str());
     return;
   }
+  ESP_LOGV(TAG, "Received UDP packet from %s, %d bytes", from.str().c_str(), n);
 
-  ESP_LOGV(TAG, "UDP Packet token data of %d bytes received", size);
-
-  if (size > MAX_DATA_LEN) {
-    ESP_LOGE(TAG, "UDP Packet too large: %d", size);
-    return;
+  if (n > 0) {
+    add_queued_request(address, token, std::move(request));
   }
-
-  network::IPAddress ip = udp.remoteIP();
-  if (udp_source_ip_.size() && std::count(udp_source_ip_.begin(), udp_source_ip_.end(), ip) == 0) {
-    ESP_LOGW(TAG, "UDP Packet wrong ip ignored %s", ip.str().c_str());
-    return;
-  }
-
-  request_data_type request;
-  request.assign(udp.data(), udp.data() + size);
-  add_queued_request(address, token, std::move(request));
 }
 
 static int copy_request(const request_data_type &request, uint8_t *data) {
@@ -118,31 +119,78 @@ void NibeGwComponent::setup() {
 
 void NibeGwComponent::dump_config() {
   ESP_LOGCONFIG(TAG, "NibeGw");
-  for (auto target = udp_targets_.begin(); target != udp_targets_.end(); target++) {
-    ESP_LOGCONFIG(TAG, " Target: %s:%d", std::get<0>(*target).str().c_str(), std::get<1>(*target));
+  for (auto &&address : udp_targets_) {
+    ESP_LOGCONFIG(TAG, " Target: %s", address.str().c_str());
   }
-  for (auto address = udp_source_ip_.begin(); address != udp_source_ip_.end(); address++) {
-    ESP_LOGCONFIG(TAG, " Source: %s", address->str().c_str());
+  for (auto &&address : udp_sources_) {
+    ESP_LOGCONFIG(TAG, " Source: %s", address.str().c_str());
   }
   ESP_LOGCONFIG(TAG, " Read Port: %d", udp_read_port_);
   ESP_LOGCONFIG(TAG, " Write Port: %d", udp_write_port_);
 }
 
+std::unique_ptr<socket::Socket> NibeGwComponent::bind_local_socket(int port) {
+  auto fd = socket::socket_ip_loop_monitored(SOCK_DGRAM, 0);
+  if (fd) {
+    // Set non-blocking
+    fd->setblocking(true);
+
+    // Bind to write port
+    socket_address address(port);
+
+    if (fd->bind((sockaddr *) &address.storage, address.len) < 0) {
+      ESP_LOGE(TAG, "Failed to bind socket to port %d, error: %d", port, errno);
+      fd.release();
+    } else {
+      ESP_LOGI(TAG, "UDP socket bound to port %d", port);
+    }
+  } else {
+    ESP_LOGE(TAG, "Failed to create socket, error: %d", errno);
+  }
+  return fd;
+}
+
 void NibeGwComponent::loop() {
-  if (network::is_connected() && !is_connected_) {
-    ESP_LOGI(TAG, "Connecting network ports.");
-    udp_read_.listen(udp_read_port_);
-    udp_write_.listen(udp_write_port_);
-    is_connected_ = true;
+  // Handle network connection state
+
+  if (network::is_connected()) {
+    if (!is_connected_) {
+      ESP_LOGI(TAG, "Connecting network ports.");
+      is_connected_ = true;
+    }
+
+    // Create and bind read socket
+    if (!udp_read_) {
+      udp_read_ = bind_local_socket(udp_read_port_);
+    }
+
+    if (!udp_write_) {
+      udp_write_ = bind_local_socket(udp_write_port_);
+    }
+  } else {
+    if (is_connected_) {
+      ESP_LOGI(TAG, "Disconnecting network ports.");
+      is_connected_ = false;
+    }
+
+    if (udp_read_) {
+      udp_read_.release();
+    }
+    if (udp_write_) {
+      udp_write_.release();
+    }
   }
 
-  if (!network::is_connected() && is_connected_) {
-    ESP_LOGI(TAG, "Disconnecting network ports.");
-    udp_read_.close();
-    udp_write_.close();
-    is_connected_ = false;
+  // Poll sockets for incoming packets
+  if (udp_read_ && udp_read_->ready()) {
+    recv_local_socket(udp_read_, MODBUS40, READ_TOKEN);
   }
 
+  if (udp_write_ && udp_write_->ready()) {
+    recv_local_socket(udp_write_, MODBUS40, WRITE_TOKEN);
+  }
+
+  // Handle high frequency loop requirement
   if (gw_->messageStillOnProgress()) {
     high_freq_.start();
   } else {

--- a/components/nibegw/NibeGwComponent.h
+++ b/components/nibegw/NibeGwComponent.h
@@ -7,30 +7,25 @@
 #include <cstdint>
 #include <map>
 
-#include "esphome.h"
 #include "esphome/core/component.h"
 #include "esphome/core/gpio.h"
+#include "esphome/core/log.h"
 #include "esphome/components/uart/uart.h"
+#include "esphome/components/network/ip_address.h"
+#include "esphome/components/network/util.h"
+#include "esphome/components/socket/socket.h"
 
 #include "NibeGw.h"
-
-#ifdef USE_ESP32
-#include <WiFi.h>
-#include "AsyncUDP.h"
-#endif
-
-#ifdef USE_ESP8266
-#include <ESP8266WiFi.h>
-#include "ESPAsyncUDP.h"
-#endif
+#include "NibeGwSockAddress.h"
 
 namespace esphome {
 namespace nibegw {
 
+using namespace std;
+
 typedef std::tuple<uint16_t, uint8_t> request_key_type;
 typedef std::vector<uint8_t> request_data_type;
 typedef std::function<request_data_type(void)> request_provider_type;
-typedef std::tuple<network::IPAddress, int> target_type;
 typedef std::function<void(const request_data_type &)> message_listener_type;
 
 class NibeGwComponent : public esphome::Component, public esphome::uart::UARTDevice {
@@ -41,10 +36,10 @@ class NibeGwComponent : public esphome::Component, public esphome::uart::UARTDev
   const int requests_queue_max = 3;
   int udp_read_port_ = 9999;
   int udp_write_port_ = 10000;
-  std::vector<network::IPAddress> udp_source_ip_;
   bool is_connected_ = false;
 
-  std::vector<target_type> udp_targets_;
+  std::vector<socket_address> udp_sources_;
+  std::vector<socket_address> udp_targets_;
   std::map<request_key_type, std::queue<request_data_type>> requests_;
   std::map<request_key_type, request_provider_type> requests_provider_;
   std::map<request_key_type, message_listener_type> message_listener_;
@@ -52,14 +47,16 @@ class NibeGwComponent : public esphome::Component, public esphome::uart::UARTDev
 
   NibeGw *gw_;
 
-  AsyncUDP udp_read_;
-  AsyncUDP udp_write_;
+  std::unique_ptr<socket::Socket> udp_read_;
+  std::unique_ptr<socket::Socket> udp_write_;
 
   void callback_msg_received(const uint8_t *data, int len);
   int callback_msg_token_received(uint16_t address, uint8_t command, uint8_t *data);
   void callback_debug(uint8_t verbose, char *data);
 
-  void token_request_cache(AsyncUDPPacket &udp, uint8_t address, uint8_t token);
+  void recv_local_socket(std::unique_ptr<socket::Socket> &fd, int address, int token);
+
+  std::unique_ptr<socket::Socket> bind_local_socket(int port);
 
  public:
   void set_read_port(int port) {
@@ -70,12 +67,11 @@ class NibeGwComponent : public esphome::Component, public esphome::uart::UARTDev
   };
 
   void add_target(const network::IPAddress &ip, int port) {
-    auto target = target_type(ip, port);
-    udp_targets_.push_back(target);
+    udp_targets_.push_back(socket_address(ip, port));
   }
 
   void add_source_ip(const network::IPAddress &ip) {
-    udp_source_ip_.push_back(ip);
+    udp_sources_.push_back(socket_address(ip, 0));
   };
 
   void set_request(int address, int token, request_data_type request) {

--- a/components/nibegw/NibeGwSockAddress.h
+++ b/components/nibegw/NibeGwSockAddress.h
@@ -1,0 +1,135 @@
+#pragma once
+
+#include <cstring>
+#include <string>
+#include <sstream>
+#include <cstdint>
+#include "esphome/components/socket/socket.h"
+
+namespace esphome {
+
+namespace socket {
+extern std::string format_sockaddr(const struct sockaddr_storage &storage);
+}
+
+namespace nibegw {
+
+// Thin wrapper around sockaddr_storage + length. Supports IPv4 matching and
+// ordering (operator<) for use in sorted containers.
+struct socket_address {
+  sockaddr_storage storage{};
+  socklen_t len = sizeof(sockaddr_storage);
+
+  socket_address() = default;
+
+  socket_address(const sockaddr *addr, socklen_t addr_len) {
+    if (addr && addr_len <= (socklen_t) sizeof(storage)) {
+      std::memcpy(&storage, addr, addr_len);
+      len = addr_len;
+    } else {
+      len = 0;
+    }
+  }
+
+  socket_address(const network::IPAddress &ip, int port) {
+    len = socket::set_sockaddr((sockaddr *) &storage, sizeof(storage), ip.str(), port);
+  }
+
+  socket_address(int port) {
+    len = socket::set_sockaddr_any((sockaddr *) &storage, sizeof(storage), port);
+  }
+
+  std::string str() const {
+    std::string s = socket::format_sockaddr(storage);
+    if (storage.ss_family == AF_INET) {
+      const auto *a = reinterpret_cast<const sockaddr_in *>(&storage);
+      uint16_t port = ntohs(a->sin_port);
+      if (port != 0) {
+        s += ":" + std::to_string(port);
+      }
+    }
+    return s;
+  }
+
+  bool valid() const {
+    return len != 0;
+  }
+
+  // Match semantics similar to compare_sockaddr(): same family, same address
+  // bytes, and if both ports are non-zero they must match. Only IPv4 is
+  // supported for now.
+  bool matches(const socket_address &other) const {
+    if (!valid() || !other.valid())
+      return false;
+
+    if (storage.ss_family != other.storage.ss_family)
+      return false;
+
+    if (storage.ss_family == AF_INET) {
+      const auto *a = reinterpret_cast<const sockaddr_in *>(&storage);
+      const auto *b = reinterpret_cast<const sockaddr_in *>(&other.storage);
+
+      if (a->sin_port != 0 && b->sin_port != 0 && a->sin_port != b->sin_port)
+        return false;
+
+      if (a->sin_addr.s_addr != b->sin_addr.s_addr)
+        return false;
+      return true;
+    }
+
+    // Other families not supported yet
+    return false;
+  }
+
+  // Strict equality: family, address bytes, and port must all match.
+  bool operator==(const socket_address &other) const {
+    if (!valid() || !other.valid())
+      return false;
+
+    if (storage.ss_family != other.storage.ss_family)
+      return false;
+
+    if (storage.ss_family == AF_INET) {
+      const auto *a = reinterpret_cast<const sockaddr_in *>(&storage);
+      const auto *b = reinterpret_cast<const sockaddr_in *>(&other.storage);
+
+      if (a->sin_port != b->sin_port)
+        return false;
+
+      if (a->sin_addr.s_addr != b->sin_addr.s_addr)
+        return false;
+      return true;
+    }
+    return false;
+  }
+
+  bool operator<(const socket_address &other) const {
+    if (storage.ss_family != other.storage.ss_family)
+      return storage.ss_family < other.storage.ss_family;
+
+    if (storage.ss_family == AF_INET) {
+      const auto *a = reinterpret_cast<const sockaddr_in *>(&storage);
+      const auto *b = reinterpret_cast<const sockaddr_in *>(&other.storage);
+
+      uint32_t aa = ntohl(a->sin_addr.s_addr);
+      uint32_t bb = ntohl(b->sin_addr.s_addr);
+      if (aa != bb)
+        return aa < bb;
+
+      uint16_t pa = ntohs(a->sin_port);
+      uint16_t pb = ntohs(b->sin_port);
+      if (pa != pb)
+        return pa < pb;
+      return false;
+    }
+
+    // Fallback: compare raw bytes
+    int cmp = std::memcmp(&storage, &other.storage, std::min((size_t) len, (size_t) other.len));
+    if (cmp != 0)
+      return cmp < 0;
+    return len < other.len;
+  }
+};
+
+}  // namespace nibegw
+}  // namespace esphome

--- a/components/nibegw/__init__.py
+++ b/components/nibegw/__init__.py
@@ -109,14 +109,6 @@ async def to_code(config):
     await cg.register_component(var, config)
     await uart.register_uart_device(var, config)
 
-    if CORE.is_esp32:
-        cg.add_build_flag("-DHARDWARE_SERIAL_WITH_PINS")
-        cg.add_library("ESP32 Async UDP", None)
-
-    if CORE.is_esp8266:
-        cg.add_build_flag("-DHARDWARE_SERIAL")
-        cg.add_library("ESPAsyncUDP", None)
-
     if udp := config.get(CONF_UDP):
         for target in udp[CONF_TARGET]:
             cg.add(var.add_target(IPAddress(str(target[CONF_TARGET_IP])), target[CONF_TARGET_PORT]))

